### PR TITLE
Introduce `_CCCL_PRAGMA` to CCCL

### LINKED ISSUE #2601: [FEA]: Compiler independent `_CCCL_PRAGMA`
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this request and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Area

General CCCL

### Is your feature request related to a problem? Please describe.

In the code base, for MSVC it is prefered to use `__pragma` over standard `_Pragma`. This leads to code duplication in many cases.

### Describe the solution you'd like

Provide compiler independent `_CCCL_PRAGMA` macro implemented as (or something similar):
```cpp
#define _CCCL_STRINGIFY2(x) #x
#define _CCCL_STRINGIFY(x) _CCCL_STRINGIFY2(x)

#ifdef _CCCL_COMPILER_MSVC
#  define _CCCL_PRAGMA(x) __pragma(x)
#else
#  define _CCCL_PRAGMA(x) _Pragma(_CCCL_STRINGIFY(x))
#endif
```
It may be used as:
```cpp
#ifdef _CCCL_COMPILER_MSVC
_CCCL_PRAGMA(message("Hi from MSVC"))
#else
_CCCL_PRAGMA(message "Hi from other compiler")
#endif
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by DejvBayer**
## Description

Closes #2601.

This PR simplifies using pragmas in macros independent on current compiler. The macro `_CCCL_PRAGMA` is defined in the `libcudacxx/include/cuda/std/__cccl/compiler.h` module.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines]().
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cub/cub/device/dispatch/dispatch_transform.cuh [hunk 1]
```diff
@@ -382,22 +382,22 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
...
-        /* need to expand into a tuple for guaranteed order of evaluation*/                                       \
-        out[idx] = poor_apply(                                                                                    \
-          [&](const InTs&... values) {                                                                            \
-            return f(values...);                                                                                  \
-          },                                                                                                      \
-          ::cuda::std::tuple<InTs...>{fetch_operand(tile_stride, smem, smem_offset, idx, aligned_ptrs)...});      \
-      }                                                                                                           \
+#  define UBLKCP_AGENT(full_tile)                                                                                    \
+    _CCCL_PRAGMA(unroll 1) /* Unroll 1 tends to improve performance, especially for smaller data types (confirmed by \
+                              benchmark) */                                                                          \
```
Comment (miscco) [inline cub/cub/device/dispatch/dispatch_transform.cuh:387]: If we are touching this, can we move the comment to the line before the pragma so it does not wrap around?

**Comment by DejvBayer** [comment]
/pre-commit.ci autofix

File: libcudacxx/include/cuda/std/__cccl/compiler.h [hunk 2]
```diff
@@ -100,4 +100,15 @@
 #  define _CCCL_CUDACC_BELOW_12_3
 #endif // defined(_CCCL_CUDACC) && _CCCL_CUDACC_VER < 1203000
 
+// Enable us to selectively silence host compiler warnings
+#define _CCCL_TOSTRING2(_STR) #_STR
+#define _CCCL_TOSTRING(_STR)  _CCCL_TOSTRING2(_STR)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__cccl/compiler.h:105]: Lets name this with a _
```suggestion
#define _CCCL_TO_STRING2(_STR) #_STR
#define _CCCL_TO_STRING(_STR)  _CCCL_TO_STRING2(_STR)
```

**Review by miscco — COMMENTED** [review]
I am a bit anxious about this tripping up some compiler bugs but lets see what CI says

File: cub/cub/device/dispatch/dispatch_transform.cuh (hunk 1 above)
Comment (DejvBayer) [reply cub/cub/device/dispatch/dispatch_transform.cuh]: Done.

**Comment by miscco** [comment]
pre-commit.ci autofix

File: libcudacxx/include/cuda/std/__cccl/compiler.h (hunk 2 above)
Comment (DejvBayer) [reply libcudacxx/include/cuda/std/__cccl/compiler.h]: Updated on all appearances.

**Review by miscco — APPROVED** [review]
I like this a lot thanks for implementing this simplification 🎉

**Pull request merged**